implementations/java-json-schema/BowtieJsonSchema.java

```diff
@@ -0,0 +1,191 @@
+                case "run" -> run(node);
+                case "stop" -> System.exit(0);
+                default -> throw new IllegalArgumentException("Unknown cmd [%s]".formatted(cmd));
+            }
+        } catch (IOException e) {
+            throw new UncheckedIOException(e);
+        }
+    }
+
+    private void start(JsonNode node) throws IOException {
+        StartRequest startRequest = objectMapper.treeToValue(node, StartRequest.class);
+        if (startRequest.version() != 1) {
+            throw new IllegalArgumentException("Unsupported IHOP version [%d]".formatted(startRequest.version()));
+        }
+
+        InputStream is = getClass().getResourceAsStream("META-INF/MANIFEST.MF");
+        var attributes = new Manifest(is).getMainAttributes();
+
+        StartResponse startResponse = new StartResponse(1, true, new Implementation(
+                "java",
+                attributes.getValue("Implementation-Name"),
+                attributes.getValue("Implementation-Version"),
+                List.of(DRAFT_2020),
+                "https://github.com/harrel56/json-schema",
+                "https://github.com/harrel56/json-schema/issues",
+                System.getProperty("os.name"),
+                System.getProperty("os.version"),
+                System.getProperty("java.vendor.version"),
+                List.of(
+                        new Link("https://harrel.dev", "Group homepage"),
+                        new Link(createMavenUrl("Implementation", attributes), "Maven Central - implementation"),
+                        new Link(createMavenUrl("Provider", attributes), "Maven Central - used JSON provider")
+                )
+        ));
+        output.println(objectMapper.writeValueAsString(startResponse));
+    }
+
+    private void dialect(JsonNode node) throws JsonProcessingException {
+        DialectRequest dialectRequest = objectMapper.treeToValue(node, DialectRequest.class);
+        if (DRAFT_2020.equals(dialectRequest.dialect())) {
+            validatorFactory.withDefaultMetaSchemaUri(DRAFT_2020);
+            output.println(objectMapper.writeValueAsString(new DialectResponse(true)));
+        } else {
+            validatorFactory.withDefaultMetaSchemaUri(null);
+            output.println(objectMapper.writeValueAsString(new DialectResponse(false)));
+        }
+    }
+
+    private void run(JsonNode node) throws JsonProcessingException {
+        RunRequest runRequest = objectMapper.treeToValue(node, RunRequest.class);
+        try {
+            if (UNSUPPORTED.containsKey(runRequest.testCase().description())) {
+                RunSkippedResponse response = new RunSkippedResponse(runRequest.seq(), true, UNSUPPORTED.get(runRequest.testCase().description()), null);
+                output.println(objectMapper.writeValueAsString(response));
+                return;
+            }
+
+            validatorFactory.withSchemaResolver(new RegistrySchemaResolver(runRequest.testCase().registry()));
```

Comment:
This can be `null` (if there's no `registry` for the test case because it references no external schemas.)

Probably here's the right place to handle that right?

Right now, the harness is failing `bowtie smoke` which is a very basic test of functionality which doesn't use a registry, but you can see it via:

```
⊙  bowtie smoke -i java-json-schema                                                                                                                                                                            julian@Airm ●
Testing 'ghcr.io/bowtie-json-schema/java-json-schema'...
  ❗ (error): allow-everything schema
  ❗ (error): allow-nothing schema

❌ some failures
```

We should really make that show the errors when they happen obviously, but you can see an equivalent by running:

```
⊙  bowtie validate -i java-json-schema <(printf '{}') <(printf 12)                                                                                                                                             julian@Airm ●
2023-05-22 11:53.40 [info     ] Will speak                     dialect=https://json-schema.org/draft/2020-12/schema
{"started": "2023-05-22T15:53:40.895747+00:00", "bowtie_version": "0.84.3.dev12+g4e542a9", "dialect": "https://json-schema.org/draft/2020-12/schema", "implementations": {"ghcr.io/bowtie-json-schema/java-json-schema": {"language": "java", "name": "json-schema", "version": "1.0.0-beta.4", "dialects": ["https://json-schema.org/draft/2020-12/schema"], "homepage": "https://github.com/harrel56/json-schema", "issues": "https://github.com/harrel56/json-schema/issues", "os": "Linux", "os_version": "6.2.13-300.fc38.aarch64", "language_version": "Temurin-17.0.7+7", "links": [{"url": "https://harrel.dev", "description": "Group homepage"}, {"url": "https://mvnrepository.com/artifact/dev.harrel/json-schema/1.0.0-beta.4", "description": "Maven Central - implementation"}, {"url": "https://mvnrepository.com/artifact/com.fasterxml.jackson.core/jackson-databind/2.15.1", "description": "Maven Central - used JSON provider"}], "image": "ghcr.io/bowtie-json-schema/java-json-schema"}}, "metadata": {}}
{"case": {"description": "bowtie validate", "schema": {}, "tests": [{"description": "1", "instance": 12, "comment": null, "valid": null}], "comment": null, "registry": null}, "seq": 1}
╭───────────────────────────────────────────────────────────────────────────────── ghcr.io/bowtie-json-schema/java-json-schema (traceback) ─────────────────────────────────────────────────────────────────────────────────╮
│                                                                                                                                                                                                                           │
│    java.lang.NullPointerException: Cannot invoke "java.util.Map.get(Object)" because "this.registry" is null                                                                                                              │
│            at BowtieJsonSchema$RegistrySchemaResolver.resolve(BowtieJsonSchema.java:149)                                                                                                                                  │
│            at dev.harrel.jsonschema.CompositeSchemaResolver.lambda$resolve$0(CompositeSchemaResolver.java:30)                                                                                                             │
│            at java.base/java.util.stream.ReferencePipeline$3$1.accept(ReferencePipeline.java:197)                                                                                                                         │
│            at java.base/java.util.Spliterators$ArraySpliterator.tryAdvance(Spliterators.java:1002)                                                                                                                        │
│            at java.base/java.util.stream.ReferencePipeline.forEachWithCancel(ReferencePipeline.java:129)                                                                                                                  │
│            at java.base/java.util.stream.AbstractPipeline.copyIntoWithCancel(AbstractPipeline.java:527)                                                                                                                   │
│            at java.base/java.util.stream.AbstractPipeline.copyInto(AbstractPipeline.java:513)                                                                                                                             │
│            at java.base/java.util.stream.AbstractPipeline.wrapAndCopyInto(AbstractPipeline.java:499)                                                                                                                      │
│            at java.base/java.util.stream.FindOps$FindOp.evaluateSequential(FindOps.java:150)                                                                                                                              │
│            at java.base/java.util.stream.AbstractPipeline.evaluate(AbstractPipeline.java:234)                                                                                                                             │
│            at java.base/java.util.stream.ReferencePipeline.findFirst(ReferencePipeline.java:647)                                                                                                                          │
│            at dev.harrel.jsonschema.CompositeSchemaResolver.resolve(CompositeSchemaResolver.java:32)                                                                                                                      │
│            at dev.harrel.jsonschema.MetaSchemaValidator.resolveExternalSchema(MetaSchemaValidator.java:44)                                                                                                                │
│            at dev.harrel.jsonschema.MetaSchemaValidator.lambda$resolveMetaSchema$2(MetaSchemaValidator.java:36)                                                                                                           │
│            at java.base/java.util.Optional.orElseGet(Optional.java:364)                                                                                                                                                   │
│            at dev.harrel.jsonschema.MetaSchemaValidator.resolveMetaSchema(MetaSchemaValidator.java:36)                                                                                                                    │
│            at dev.harrel.jsonschema.MetaSchemaValidator.validateMetaSchema(MetaSchemaValidator.java:24)                                                                                                                   │
│            at dev.harrel.jsonschema.JsonParser.parseRootSchema(JsonParser.java:47)                                                                                                                                        │
│            at dev.harrel.jsonschema.Validator.registerSchema(Validator.java:32)                                                                                                                                           │
│            at dev.harrel.jsonschema.ValidatorFactory.validate(ValidatorFactory.java:78)                                                                                                                                   │
│            at dev.harrel.jsonschema.ValidatorFactory.validate(ValidatorFactory.java:61)                                                                                                                                   │
│            at BowtieJsonSchema.lambda$run$0(BowtieJsonSchema.java:113)                                                                                                                                                    │
│            at java.base/java.util.stream.ReferencePipeline$3$1.accept(ReferencePipeline.java:197)                                                                                                                         │
│            at java.base/java.util.ArrayList$ArrayListSpliterator.forEachRemaining(ArrayList.java:1625)                                                                                                                    │
│            at java.base/java.util.stream.AbstractPipeline.copyInto(AbstractPipeline.java:509)                                                                                                                             │
│            at java.base/java.util.stream.AbstractPipeline.wrapAndCopyInto(AbstractPipeline.java:499)                                                                                                                      │
│            at java.base/java.util.stream.AbstractPipeline.evaluate(AbstractPipeline.java:575)                                                                                                                             │
│            at java.base/java.util.stream.AbstractPipeline.evaluateToArrayNode(AbstractPipeline.java:260)                                                                                                                  │
│            at java.base/java.util.stream.ReferencePipeline.toArray(ReferencePipeline.java:616)                                                                                                                            │
│            at java.base/java.util.stream.ReferencePipeline.toArray(ReferencePipeline.java:622)                                                                                                                            │
│            at java.base/java.util.stream.ReferencePipeline.toList(ReferencePipeline.java:627)                                                                                                                             │
│            at BowtieJsonSchema.run(BowtieJsonSchema.java:115)                                                                                                                                                             │
│            at BowtieJsonSchema.handle(BowtieJsonSchema.java:54)                                                                                                                                                           │
│            at java.base/java.util.Iterator.forEachRemaining(Iterator.java:133)                                                                                                                                            │
│            at java.base/java.util.Spliterators$IteratorSpliterator.forEachRemaining(Spliterators.java:1845)                                                                                                               │
│            at java.base/java.util.stream.ReferencePipeline$Head.forEach(ReferencePipeline.java:762)                                                                                                                       │
│            at BowtieJsonSchema.run(BowtieJsonSchema.java:44)                                                                                                                                                              │
│            at BowtieJsonSchema.main(BowtieJsonSchema.java:36)                                                                                                                                                             │
│                                                                                                                                                                                                                           │
╰───────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────╯
2023-05-22 11:53.40 [error    ]                                [ghcr.io/bowtie-json-schema/java-json-schema] case=bowtie validate message=Cannot invoke "java.util.Map.get(Object)" because "this.registry" is null schema={} seq=1
{"implementation": "ghcr.io/bowtie-json-schema/java-json-schema", "seq": 1, "context": {"message": "Cannot invoke \"java.util.Map.get(Object)\" because \"this.registry\" is null", "traceback": "java.lang.NullPointerException: Cannot invoke \"java.util.Map.get(Object)\" because \"this.registry\" is null\n\tat BowtieJsonSchema$RegistrySchemaResolver.resolve(BowtieJsonSchema.java:149)\n\tat dev.harrel.jsonschema.CompositeSchemaResolver.lambda$resolve$0(CompositeSchemaResolver.java:30)\n\tat java.base/java.util.stream.ReferencePipeline$3$1.accept(ReferencePipeline.java:197)\n\tat java.base/java.util.Spliterators$ArraySpliterator.tryAdvance(Spliterators.java:1002)\n\tat java.base/java.util.stream.ReferencePipeline.forEachWithCancel(ReferencePipeline.java:129)\n\tat java.base/java.util.stream.AbstractPipeline.copyIntoWithCancel(AbstractPipeline.java:527)\n\tat java.base/java.util.stream.AbstractPipeline.copyInto(AbstractPipeline.java:513)\n\tat java.base/java.util.stream.AbstractPipeline.wrapAndCopyInto(AbstractPipeline.java:499)\n\tat java.base/java.util.stream.FindOps$FindOp.evaluateSequential(FindOps.java:150)\n\tat java.base/java.util.stream.AbstractPipeline.evaluate(AbstractPipeline.java:234)\n\tat java.base/java.util.stream.ReferencePipeline.findFirst(ReferencePipeline.java:647)\n\tat dev.harrel.jsonschema.CompositeSchemaResolver.resolve(CompositeSchemaResolver.java:32)\n\tat dev.harrel.jsonschema.MetaSchemaValidator.resolveExternalSchema(MetaSchemaValidator.java:44)\n\tat dev.harrel.jsonschema.MetaSchemaValidator.lambda$resolveMetaSchema$2(MetaSchemaValidator.java:36)\n\tat java.base/java.util.Optional.orElseGet(Optional.java:364)\n\tat dev.harrel.jsonschema.MetaSchemaValidator.resolveMetaSchema(MetaSchemaValidator.java:36)\n\tat dev.harrel.jsonschema.MetaSchemaValidator.validateMetaSchema(MetaSchemaValidator.java:24)\n\tat dev.harrel.jsonschema.JsonParser.parseRootSchema(JsonParser.java:47)\n\tat dev.harrel.jsonschema.Validator.registerSchema(Validator.java:32)\n\tat dev.harrel.jsonschema.ValidatorFactory.validate(ValidatorFactory.java:78)\n\tat dev.harrel.jsonschema.ValidatorFactory.validate(ValidatorFactory.java:61)\n\tat BowtieJsonSchema.lambda$run$0(BowtieJsonSchema.java:113)\n\tat java.base/java.util.stream.ReferencePipeline$3$1.accept(ReferencePipeline.java:197)\n\tat java.base/java.util.ArrayList$ArrayListSpliterator.forEachRemaining(ArrayList.java:1625)\n\tat java.base/java.util.stream.AbstractPipeline.copyInto(AbstractPipeline.java:509)\n\tat java.base/java.util.stream.AbstractPipeline.wrapAndCopyInto(AbstractPipeline.java:499)\n\tat java.base/java.util.stream.AbstractPipeline.evaluate(AbstractPipeline.java:575)\n\tat java.base/java.util.stream.AbstractPipeline.evaluateToArrayNode(AbstractPipeline.java:260)\n\tat java.base/java.util.stream.ReferencePipeline.toArray(ReferencePipeline.java:616)\n\tat java.base/java.util.stream.ReferencePipeline.toArray(ReferencePipeline.java:622)\n\tat java.base/java.util.stream.ReferencePipeline.toList(ReferencePipeline.java:627)\n\tat BowtieJsonSchema.run(BowtieJsonSchema.java:115)\n\tat BowtieJsonSchema.handle(BowtieJsonSchema.java:54)\n\tat java.base/java.util.Iterator.forEachRemaining(Iterator.java:133)\n\tat java.base/java.util.Spliterators$IteratorSpliterator.forEachRemaining(Spliterators.java:1845)\n\tat java.base/java.util.stream.ReferencePipeline$Head.forEach(ReferencePipeline.java:762)\n\tat BowtieJsonSchema.run(BowtieJsonSchema.java:44)\n\tat BowtieJsonSchema.main(BowtieJsonSchema.java:36)\n"}, "caught": true}
2023-05-22 11:53.40 [info     ] Finished                       count=1
{"did_fail_fast": false}
```

Reply:
This is a good catch, but honestly I overlooked this null pointer exception - I was pretty sure that smoke tests failed because of lack of harness image in registry? I saw that publish phase in pipeline was omitted. Nevertheless, I fixed that null issue, thanks!

Reply:
Still, I see different error log than you posted:
```
Run bowtie smoke -i "java-json-schema:latest"
  bowtie smoke -i "java-json-schema:latest"
  shell: /usr/bin/bash -e {0}
❗ (error): 'ghcr.io/bowtie-json-schema/java-json-schema:latest' is not a known Bowtie implementation.
```
Am I missing something?

Reply:
This is a subtle thing which I think is slightly confusing to new implementers (feedback / docs suggestions welcome) but is really helpful for users of bowtie -- so basically if you say `bowtie smoke -i foo` bowtie assumes you mean the `foo` image that is the "official built bowtie image", so it will go fetch that (by retrieving the image `ghcr.io/bowtie-json-schema/foo` out of the GitHub container registry).

But here you don't mean that! You have built an image locally and want bowtie to smoke test that one.

So if you called your image just plain `podman build -t java-json-schema` or whatever, you need to be a bit more specific that you want the local image.

Specifically, `bowtie smoke -i "localhost/java-json-schema"` I *think* should work.

Reply:
Yeah, I think the behavior you described is pretty clear, although I haven't used `bowtie smoke` but I think the rules for `bowtie suite` are the same?
But why the pipeline is still failing? Shouldn't it use the official registry image for smoke tests? It should get built and published beforehand, right?

Reply:
They're the same yeah!

> Shouldn't it use the official registry image for smoke tests? It should get built and published beforehand, right?

GitHub actions runners have some weird behavior I haven't been able to figure out yet, which was that small bug I mentioned all the way at the beginning.

In theory I wanted bowtie smoke to run *before* pushing, with the point being to catch images that are completely broken, and therefore not push a broken one overwriting a working one.

So in theory I wanted to smoke the local image *before* pushing. However, GH actions runners seem to do this really weird thing where from step to step they clear their local container image cache such that when you try to run an image you just built, you can't, and I don't understand why, it doesn't seem documented.

So right now we *do* smoke test the pushed image and so the build is still failing trying to access a not-yet-pushed image for you.

I think for now the right thing to do is probably to merge anyhow, the image will then get built and future builds will pass. But gonna take one more look tomorrow now that the other changes are addressed.

Reply:
Sounds reasonable - gonna wait for updates then. If there is something more required to do on my end, let me know